api/types/common/common.go

```diff
@@ -0,0 +1,433 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"regexp"
+	"time"
+
+	commonpb "github.com/gravitational/teleport/api/gen/proto/go/common/v1"
+	"github.com/gravitational/teleport/api/utils"
+	"github.com/gravitational/trace"
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+)
+
+// Builder is an interface that can build a resource object.
+type Builder[T any] interface {
+	// Build will build a resource object.
+	Build() (T, error)
+}
+
+// ResourceHeaderBuilder is a builder that can be attached to another resource builder.
+// This will give the resource header build options to the given builder.
+type ResourceHeaderBuilder[T Builder[V], V any] struct {
+	parentBuilder T
+
+	kind    string
+	subKind string
+	version string
+	*MetadataBuilder[T, V]
+
+	resourceHeader *ResourceHeader
+}
+
+// NewResourceHeaderBuilder creates a new resource header builder with the given parent builder.
+func NewResourceHeaderBuilder[T Builder[V], V any](parent T) *ResourceHeaderBuilder[T, V] {
+	return &ResourceHeaderBuilder[T, V]{
+		parentBuilder: parent,
+		MetadataBuilder: &MetadataBuilder[T, V]{
+			parentBuilder: parent,
+		},
+	}
+}
+
+// Kind sets the kind for the resource header.
+func (r *ResourceHeaderBuilder[T, _]) Kind(kind string) T {
+	r.kind = kind
+	return r.parentBuilder
+}
+
+// SubKind sets the sub kind for the resource header.
+func (r *ResourceHeaderBuilder[T, _]) SubKind(subKind string) T {
+	r.subKind = subKind
+	return r.parentBuilder
+}
+
+// Version sets the version for the resource header.
+func (r *ResourceHeaderBuilder[T, _]) Version(version string) T {
```

Comment:
I'm debating whether `Version` has a point with this approach. Thoughts?

Reply:
It looks like, given https://github.com/gravitational/teleport/pull/27153, that this could still be useful. I'll go ahead and resolve this.